skimage/filters/thresholding.py

```diff
@@ -713,3 +716,283 @@ def threshold_triangle(image, nbins=256):
+        deviation values.
+
+    References
+    ----------
+    .. [1] F. Shafait, D. Keysers, and T. M. Breuel, "Efficient
+           implementation of local adaptive thresholding techniques
+           using integral images." in Document Recognition and
+           Retrieval XV, (San Jose, USA), Jan. 2008.
+    """
+    from skimage.transform.integral import integral_image
+
+    if w == 1 or w % 2 == 0:
+        raise ValueError(
+            "Window size w = %s must be odd and greater than 1." % (w))
+    I = integral_image(image)  # Integral Image.
+
+    # Pad left and top of Integral image with zeros
+    I = np.vstack((np.zeros((1, I.shape[1]), I.dtype), I))
+    I = np.hstack((np.zeros((I.shape[0], 1), I.dtype), I))
+
+    kern = np.zeros((w + 1, w + 1))
+    kern[0, 0], kern[-1, -1] = 1, 1
+    kern[[0, -1], [-1, 0]] = -1
+    # w2 holds total of pixels in window for each pixel (usually w x w).
+    w2 = ndi.convolve(
+        np.ones(image.shape, np.float), np.ones((w, w)), mode='constant')
+
+    m = ndi.convolve(I, kern, mode='nearest')[:-1, :-1] / w2
+    g = image.astype(np.float)
+    g2 = g ** 2.
+    m2 = m ** 2.
+    sum_g2 = ndi.convolve(g2, np.ones((w, w)), mode='constant')
+    sum_m2 = w2 * m2
+    s2 = (sum_g2 - sum_m2) / w2
+    s = np.sqrt(s2)
+    return m, s
+
+
+def threshold_niblack(image, window_size=15, k=0.2, offset=0):
+    """Applies Niblack local threshold to an array.
+
+    A threshold T is calculated for every pixel in the image using the
+    following formula:
+
+    T = m(x,y) - k * s(x,y) - offset
+
+    where m(x,y) and s(x,y) are the mean and standard deviation of
+    pixel (x,y) neighborhood defined by a rectangular window with size w
+    times w centered around the pixel. k is a configurable parameter
+    that weights the effect of standard deviation.
+
+    Parameters
+    ----------
+    image: (N, M) ndarray
+        Input image. Only 2D grayscale images allowed.
+    window_size : int, optional
+        Odd size of pixel neighborhood window (e.g. 3, 5, 7,
+        ..., 21, ...).
+    k : float, optional
+        Value of parameter k in threshold formula.
+    offset : float, optional
+        Constant subtracted from obtained local thresholds.
+
+    Returns
+    -------
+    threshold : (N, M) ndarray
+        Local threshold values. All pixels with an intensity higher than
+        this value are assumed to be foreground.
+
+    References
+    ----------
+    .. [1] Niblack, W (1986), An introduction to Digital Image
+           Processing, Prentice-Hall.
+
+    Examples
+    --------
+    >>> from skimage import data
+    >>> image = data.page()
+    >>> binary_image = threshold_niblack(image, window_size=7, k=0.1)
+    """
+    if len(image.shape) != 2:
+        raise ValueError("Image is not 2D grayscale.")
+
+    m, s = _mean_std(image, window_size)
+    t = m - k * s
+    return t - offset
+
+
+def threshold_sauvola(image, window_size=15, k=0.2, r=128., offset=0):
```

Comment:
A default value ``r=128.`` supposes an image of uint8. I'd rather use a value between 0 and 1, and modify it inside the function depending on the data type.